beacon/light/committee_chain.go

```diff
@@ -0,0 +1,555 @@
+			}
+			break
+		}
+		if batch == nil {
+			batch = s.db.NewBatch()
+		}
+		s.rollback(batch, s.updates.AfterLast)
+	}
+	if batch != nil {
+		if err := batch.Write(); err != nil {
+			log.Error("Error writing batch into chain database", "error", err)
+		}
+	}
+	return s
+}
+
+// Reset resets the committee chain.
+func (s *CommitteeChain) Reset() {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	batch := s.db.NewBatch()
+	s.rollback(batch, 0)
+	if err := batch.Write(); err != nil {
+		log.Error("Error writing batch into chain database", "error", err)
+	}
+}
+
+// AddFixedRoot sets a fixed committee root at the given period.
+// Note that the period where the first committee is added has to have a fixed
+// root which can either come from a CheckpointData or a trusted source.
+func (s *CommitteeChain) AddFixedRoot(period uint64, root common.Hash) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	batch := s.db.NewBatch()
+	oldRoot := s.getCommitteeRoot(period)
+	if !s.fixedRoots.CanExpand(period) {
+		if root != oldRoot {
+			return ErrInvalidPeriod
+		}
+		for p := s.fixedRoots.AfterLast; p <= period; p++ {
+			s.fixedRoots.add(batch, p, s.getCommitteeRoot(p))
+		}
+	}
+	if oldRoot != (common.Hash{}) && (oldRoot != root) {
+		// existing old root was different, we have to reorg the chain
+		s.rollback(batch, period)
+	}
+	s.fixedRoots.add(batch, period, root)
+	if err := batch.Write(); err != nil {
+		log.Error("Error writing batch into chain database", "error", err)
+		return err
+	}
+	return nil
+}
+
+// DeleteFixedRootsFrom deletes fixed roots starting from the given period.
+// It also maintains chain consistency, meaning that it also deletes updates and
+// committees if they are no longer supported by a valid update chain.
+func (s *CommitteeChain) DeleteFixedRootsFrom(period uint64) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if period >= s.fixedRoots.AfterLast {
+		return nil
+	}
+	batch := s.db.NewBatch()
+	s.fixedRoots.deleteFrom(batch, period)
+	if s.updates.IsEmpty() || period <= s.updates.First {
+		s.updates.deleteFrom(batch, period)
+		s.deleteCommitteesFrom(batch, period)
+	} else {
+		fromPeriod := s.updates.AfterLast + 1
+		if period > fromPeriod {
+			fromPeriod = period
+		}
+		s.deleteCommitteesFrom(batch, fromPeriod)
+	}
+	if err := batch.Write(); err != nil {
+		log.Error("Error writing batch into chain database", "error", err)
+		return err
+	}
+	return nil
+}
+
+// deleteCommitteesFrom deletes committees starting from the given period.
+func (s *CommitteeChain) deleteCommitteesFrom(batch ethdb.Batch, period uint64) {
+	deleted := s.committees.deleteFrom(batch, period)
+	for period := deleted.First; period < deleted.AfterLast; period++ {
+		s.syncCommitteeCache.Remove(period)
+	}
+}
+
+// GetCommittee returns the committee at the given period.
+func (s *CommitteeChain) GetCommittee(period uint64) *types.SerializedSyncCommittee {
+	return s.committees.get(period)
+}
```

Comment:
For both `GetCommittee` and `GetUpdate`, you omit the mutex lock. Intended?

Reply:
Yes, they are thread safe and are used both while the mutex is locked and unlocked (no need for locked and unlocked version of the same fn). Added comments about it.